Rework unit loading to take into account all aliases

keszybz opened the pull request:
~~Only the last 7 patches are new, the rest is the same as in  #13096. I'll keep rebasing this until the other one is merged.~~

For #11972.

keszybz commented:
CentOS CI (Arch in KVM) is failing with:
```
$ systemctl enable rpcbind && systemctl start rpcbind
Failed to start rpcbind.service: Unit rpcbind.service not found.
```
It seems arch might be doing something special with how rpcbind is defined... No idea.

mrc0mmand commented:
@keszybz I tweaked a little the *sanity boot test* in the CentOS 7 job (by adding `systemd.log_target=console`), maybe it'll help you with debugging https://ci.centos.org/job/systemd-pr-build/7517/artifact//systemd-centos-ci/artifacts_1U2hyV/bootstrap-logs-upstream.6iv/sanity-boot-check.log

mrc0mmand commented:
According to the CentOS CI, systemd doesn't detect newly created units without calling `systemctl daemon-reload`.

TEST-03-JOBS:

```
+ cat   # cat <<EOF >  /run/systemd/system/wait2.service...
+ cat   # cat <<EOF >  /run/systemd/system/wait5fail.service...
++ date -u +%s
+ START_SEC=1563640040
+ systemctl start --wait wait2.service
Failed to start wait2.service: Unit wait2.service not found.
[FAILED] Failed to start Testsuite service.
See 'systemctl status testsuite.service' for details.
         Starting End the test...
```

TEST-12-ISSUE-3171

```bash
...
U=/run/systemd/system/test.socket
cat <<'EOL' >$U
...
systemctl start test.socket
...
```

```
Jul 20 18:29:16 systemd-testsuite test-socket-group.sh[32]: Failed to start test.socket: Unit test.socket not found.
Jul 20 18:29:16 systemd-testsuite systemd[1]: test.socket: Collecting.
Jul 20 18:29:16 systemd-testsuite systemd[1]: test.socket: Failed to send unit remove signal for test.socket: Connection reset by peer
Jul 20 18:29:16 systemd-testsuite systemd[1]: Bus private-bus-connection: changing state CLOSING → CLOSED
Jul 20 18:29:16 systemd-testsuite systemd[1]: Got disconnect on private connection.
Jul 20 18:29:16 systemd-testsuite systemd[1]: Received SIGCHLD from PID 32 (test-socket-gro).
Jul 20 18:29:16 systemd-testsuite systemd[1]: Child 32 (test-socket-gro) died (code=exited, status=5/NOTINSTALLED)
Jul 20 18:29:16 systemd-testsuite systemd[1]: testsuite.service: Child 32 belongs to testsuite.service.
Jul 20 18:29:16 systemd-testsuite systemd[1]: testsuite.service: Main process exited, code=exited, status=5/NOTINSTALLED
Jul 20 18:29:16 systemd-testsuite systemd[1]: testsuite.service: Failed with result 'exit-code'.
Jul 20 18:29:16 systemd-testsuite systemd[1]: testsuite.service: Changed start -> failed
```

etc.

keszybz commented:
> According to the CentOS CI, systemd doesn't detect newly created units without calling systemctl daemon-reload.

Duh, a type (`errno != -ENOENT`) :(. Thanks for help diagnosing this.

I'm now happy with the latest version. It seems pretty clean. There's still some minor details to work out, but in general this should be mergeable.

poettering commented on src/shared/unit-file.c:
```diff
@@ -21,3 +22,60 @@ bool unit_type_may_template(UnitType type) {
+                                        "%s points to \"%s\" which is not a valid unit name: %m", filename, dst);
+        if (r2 == UNIT_NAME_INSTANCE)
+                return log_notice_errno(SYNTHETIC_ERRNO(EXDEV),
+                                        "%s: unit symlink target \"%s\" has instance name, rejecting.", filename, dst);
```
hmm, this is a new limitation that I am not sure I agree with? i mean, we previously said as long as the instance name matches you can have aliases, and that when searching for a file we'd always look for the name with the instance first, and without it as fallback. Why take this away? Are you sure this is nowhere used?

keszybz replied:
OK, I'll remove this limitation. (BTW, thanks, this is the kind of comments I was looking for).

keszybz replied:
Oh, I remember why this is a problem. Normally aliases are functions (in the mathematical sense, i.e. there can by only one y for any given x). This is nicely enforced by the fact that a symlink can only point one way, and once we find the symlink, we don't look for any fragments with lower priority. But once we allow symlinks for aliases, we can have `a@foo.service → b@foo.service` and at the same time `a@.service → c@.service`, while both `b@.service` and `c@service` can have fragments on disk. So are now `a@.service`, `b@.service`, `c@.service` now all aliases? If yes, do we pick the fragment for `b@.service` or `c@.service` ?

keszybz replied:
Actually, this can be even simpler: `a@.service` may exist on disk, and `b@foo.service` may be a symlink to `b@foo.service`. Now we have two candidate fragments again.

Maybe we should allow this, check that the instance alias matches the main alias, but warn?

poettering replied:
I'd say the rule is that if `a@foo.service` → `b@foo.service` exists that `a@.service` is never itself read (but the drop-ins for that are.)

i.e. in your example, I think we should read:

* `b@foo.service`
* `a@foo.service.d/*.conf`
* `b@foo.service.d/*.conf`

But not:

* `c@foo.service`
* 'c@foo.service.d/*`

i.e. we only collect the drop-ins on the symlink chain we actually follow, and we stop searching as soon as we found the first unit file that is not a symlink. that means we'd never be tempted to resolve `a@.service` at all, since we already found a unit file by traversing `a@foo.service`.

keszybz replied:
Dunno, the fact that we read `a@.service.d/*.conf` but not `a@.service` seems very arbitrary. Before, it didn't matter if a setting was defined in the fragment or in the drop-in (apart from ordering issues). But now the dropins start living a life independent of the main unit.

keszybz replied:
Hmm, I can't convince myself that this is OK.
We either consider the instanced unit `b@foo` independent of the template `a@.service`, or not.
If the first, then we shouldn't load `a@.service` or `a@.service.d`. If the second, we should load both.

poettering commented on src/shared/unit-file.c:
```diff
@@ -21,3 +22,60 @@ bool unit_type_may_template(UnitType type) {
+        dst = basename(target);
+
+        /* Reject cases where the alias or the target unit have instance names. */
+        r = unit_name_to_instance(src, NULL);
```
use unit_name_classify()?

keszybz replied:
Right.

poettering replied:
maybe call this return value src_name_type rather than r? after all you keep it for longer

poettering commented on src/shared/unit-file.c:
```diff
@@ -21,3 +22,60 @@ bool unit_type_may_template(UnitType type) {
+                                        "%s: templates not allowed for %s units, rejecting.",
+                                        filename, unit_type_to_string(src_type));
+
+        r2 = unit_name_to_instance(dst, NULL);
```
r2 → dst_name_type?

poettering commented on src/analyze/analyze.c:
```diff
@@ -1545,6 +1545,53 @@ static int get_or_set_log_target(int argc, char *argv[], void *userdata) {
+        char **s;
+        STRV_FOREACH(s, strv)
+                if (strv_fnmatch_or_empty(patterns, *s, flags))
+                    return true;
```
indentation

poettering commented on src/shared/unit-file.c:
```diff
@@ -118,8 +118,33 @@ static int unit_ids_map_get(
         return -ELOOP;
 }
 
+static bool lp_mtime_good(const LookupPaths *lp, usec_t mtime) {
```
why abbrev the function name, let's just call this looup_paths_mtime_good?

abbreviation arg names is fine, but function names? just makes it hard to read backtraces

poettering commented on src/shared/unit-file.c:
```diff
@@ -79,3 +86,299 @@ int unit_validate_alias_symlink_and_warn(const char *filename, const char *targe
+                        return 0;
+                }
+
+                unit_name = t;
```
noone sets `id` here?

mrc0mmand commented:
`TEST-15-DROPIN` is failing:

```
Jul 26 14:59:46 systemd-testsuite test-dropin.sh[29]: + check_ko a1 Wants y.service
Jul 26 14:59:46 systemd-testsuite test-dropin.sh[29]: + check_ok a1 Wants y.service
Jul 26 14:59:46 systemd-testsuite test-dropin.sh[29]: + '[' 3 -eq 3 ']'
Jul 26 14:59:46 systemd-testsuite test-dropin.sh[29]: ++ systemctl show --value -p Wants a1
Jul 26 14:59:46 systemd-testsuite test-dropin.sh[29]: + x='x.service y.service'
Jul 26 14:59:46 systemd-testsuite test-dropin.sh[29]: + case "$x" in
Jul 26 14:59:46 systemd-testsuite test-dropin.sh[29]: + return 0
Jul 26 14:59:46 systemd-testsuite systemd[1]: testsuite.service: Main process exited, code=exited, status=1/FAILURE
Jul 26 14:59:46 systemd-testsuite systemd[1]: testsuite.service: Failed with result 'exit-code'.
Jul 26 14:59:46 systemd-testsuite systemd[1]: Failed to start Testsuite service.
```

In this part of the code:
https://github.com/systemd/systemd/blob/6a0899105c40da2c381fe55c0f93f6e4ae9770b0/test/TEST-15-DROPIN/test-dropin.sh#L149-L150

With appropriate description here:
https://github.com/systemd/systemd/blob/6a0899105c40da2c381fe55c0f93f6e4ae9770b0/test/TEST-15-DROPIN/test-dropin.sh#L132-L139

poettering commented on src/shared/unit-file.c:
```diff
@@ -37,14 +38,11 @@ int unit_validate_alias_symlink_and_warn(const char *filename, const char *targe
-                                        "%s: unit symlink has instance name, rejecting.", filename);
 
         src_type = unit_name_to_type(src);
         assert(src_type >= 0); /* unit_name_classify() checked the suffix already */
```
comment needs update

poettering commented on src/shared/unit-file.c:
```diff
@@ -53,19 +51,35 @@ int unit_validate_alias_symlink_and_warn(const char *filename, const char *targe
+                if (src_name_type != UNIT_NAME_INSTANCE)
+                        return log_notice_errno(SYNTHETIC_ERRNO(EXDEV),
+                                                "%s: unit symlink target \"%s\" has instance name, rejecting.",
+                                                filename, dst);
```
this error doesn't look right... also, we could probably drop this, if we swap the order of the next if block with this one

poettering commented on src/systemctl/systemctl.c:
```diff
@@ -165,12 +166,18 @@ static bool arg_jobs_before = false;
 
+/* This is a global cache that will be constructed on first use. */
+Hashmap *arg_id_map = NULL;
+Hashmap *arg_name_map = NULL;
```
why not static?

poettering replied:
why prefixed with arg_? This isn't precisely something that is derived directly from the command line or an env var or so?

keszybz replied:
We had a general rule that global args are prefixed with `arg_`, no? I think you said so yourself.

poettering replied:
Well, we had a rule of avoiding them, as much as we can except if they inherently are a per-process concept, i.e. something derived directly from the process' comand line or environment block... Those two are summarized as "arguments", and hence get the "arg_" prefix... However, all other global variables should get something else, and be avoided unless we have a really good reason to have them...

https://systemd.io/CODING_STYLE#command-line
https://systemd.io/CODING_STYLE#code-organization-and-semantics

keszybz commented:
Another update. I added an extensive test to verify the unit loading behaviour. It is first added in a version that passes before are changes, and then when the code is changed, the test is updated again. Somewhat surprisingly, the results for the old code don't make much sense.

The result is not quite as you describe in https://github.com/systemd/systemd/pull/13119#discussion_r308145418 (difference is crossed out):

1. `b@foo.service`
2. ~~`a@.service.d/*.conf`~~
3. `a@foo.service.d/*.conf`
4. `b@.service.d/*.conf`
5. `b@foo.service.d/*.conf`

But not:

* `a@.service` (i.e. that this is a symlink is irrelevant to us, we already found a unit with the `a@foo.service` chain)
* `c@.service`
* `c@foo.service`
* `c@foo.service.d/*.conf`
* `c@.service.d/*.conf`

My thinking is: if we aliased a unit "away" from some template, this unit does not use this template, and we load neither the template fragment, nor any dropins declared for the template. OTOH, we load both the fragment and the dropins for the new template. Of course, we load all dropins for the instance, but all names. This means that declaring something directly in the template fragment and in the template dropin is equivalent.

I corrected some minor bugs on the way and cleaned up the debugging stmts and FIXMEs. (One remains, but that's something I'm leaving for later). PTAL.

poettering commented:
hmm, lots's of CI fialures?

keszybz commented:
It seems to be timeouts, e.g. on systemd-update-hwdb.service. I restarted the tests.

keszybz commented:
`Failed to transfer image: Connection timed out` — yeah, that's not likely to be related.

keszybz commented:
```
# fix paths in manpages; manually check the remaining /usr occurrences
# occasionally, with filtering out paths which are known to be in /usr:
# grep -r /usr debian/install/deb/usr/share/man/|egrep -v '/usr/local|os.*release|factory|zoneinfo|tmpfiles|kernel|foo|machines|sysctl|dbus|include|binfmt'
find debian/install/deb/usr/share/man/ -type f | xargs sed -ri 's_/usr(/lib/systemd/system|/lib/systemd/network|/lib/udev|/lib[^/]|/lib/[^a-z])_\1_g'
find: ‘debian/install/deb/usr/share/man/’: No such file or directory
sed: no input files
```

evverx commented:
@keszybz @poettering could you at least cc @ddstreet when something like this happens? It looks like that particular issue can be fixed by passing `-Dman=true` to `meson` in `debian/rules`.

Another option (especially after https://github.com/systemd/systemd/pull/13221) would be to stop pretending anyone looks at Ubuntu CI when PRs are opened and simply turn it off.

evverx commented:
@keszybz I'm not sure how you restarted Ubuntu CI but it seems the tests were run against the "master" branch of the Debian package where @mbiebl reverted the commit where `-Dman=true` was passed to `meson` in https://salsa.debian.org/systemd-team/systemd/commit/ccf7a5dc83928beac5bdadfabef911e6082131e2. The test should be run against the "experimental" branch. That commit hasn't been reverted there.

keszybz commented:
Oh, I just use the `./retry-gh-systemd-test` that I got from @martinpitt ages ago. It's possible that it is somehow outdated...

evverx commented:
@keszybz could you add the script to the repository (without the api keys) so that we could more or less keep it up to date? By the way, we discussed this fragile scheme (which we also rely on on Semaphore) in https://github.com/systemd/systemd/pull/12980.

ddstreet commented:
> @keszybz @poettering could you at least cc @ddstreet when something like this happens?

Yes please do - I'm happy to look at any Ubuntu CI failures/problems.  I'm working on getting the Ubuntu CI more stable/reliable, as well.

buczek commented:
Did e8630e6952 ("pid1: use a cache for all unit aliases") change the semantic of chains like `/etc/systemd/system/default.target` -> `/etc/local/systemd_system_default.target` -> ` /lib/systemd/system/graphical.target` ? It looks like in v242 such a chain would be interpreted as an alias and in v260.2 it would be interpreted as a "linked unit file" with the result that dependencies of graphical.target are ignored.
